src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -28,99 +28,215 @@
+     * This method may only be called in "owner" thread.
+     *
+     * @throws IllegalStateException if this scope is already closed or if this is
+     *                               a root scope and there is/are still active child
+     *                               scope(s).
+     */
+    abstract void close();
 
-    public MemoryScope(Object ref, Runnable cleanupAction) {
-        this.ref = ref;
-        this.cleanupAction = cleanupAction;
-    }
+    /**
+     * Duplicates this scope and {@link #close() closes} it. If this is a root scope,
+     * new root scope is returned. If this is a child scope, new child scope is returned.
+     * This method may only be called in "owner" thread.
+     * The returned instance may be published unsafely to and used in any thread, but methods that explicitly state that
+     * they may only be called in "owner" thread, must strictly be called in single thread that has been selected to be the
+     * "owner" thread.
+     *
+     * @return a duplicate of this scope
+     * @throws IllegalStateException if this scope is already closed or if this is
+     *                               a root scope and there is/are still active child
+     *                               scope(s).
+     */
+    abstract MemoryScope dup();
 
     /**
-     * This method performs a full, thread-safe liveness check; can be used outside confinement thread.
+     * This method may be called in any thread.
+     *
+     * @return {@code true} if this scope is not closed yet.
      */
     final boolean isAliveThreadSafe() {
-        return ((int)COUNT_HANDLE.getVolatile(this)) != CLOSED;
+        return ((int) STATE.getVolatile(this)) < STATE_CLOSED;
     }
 
     /**
-     * This method performs a quick liveness check; must be called from the confinement thread.
+     * Checks that this scope is still alive.
+     * This method may only be called in "owner" thread.
+     *
+     * @throws IllegalStateException if this scope is already closed
      */
     final void checkAliveConfined() {
-        if (activeCount == CLOSED) {
-            throw new IllegalStateException("Segment is not alive");
+        if (state == STATE_CLOSED) {
+            throw new IllegalStateException("This scope is already closed");
         }
     }
 
-    MemoryScope acquire() {
-        int value;
-        do {
-            value = (int)COUNT_HANDLE.getVolatile(this);
-            if (value == CLOSED) {
-                //segment is not alive!
-                throw new IllegalStateException("Segment is not alive");
-            } else if (value == MAX_ACQUIRE) {
-                //overflow
-                throw new IllegalStateException("Segment acquire limit exceeded");
-            }
-        } while (!COUNT_HANDLE.compareAndSet(this, value, value + 1));
-        return new MemoryScope(ref, this::release);
-    }
+    private static final class Root extends MemoryScope {
+        private final LongAdder acquires;
```

Comment:
Did you consider collapsing into just one `LongAdder` using `increment` (for acquiring), `decrement` (for releasing), and `sum` (checking a zero sum for closing)?  Perhaps it was discussed already, i lost track of all the discussion.

Reply:
That would not work. LongAdder.sum() is not an atomic operation. Consider the following scenario:
- precondition: adder.sum() == 0
- thread A1 successfully acquires child segment: adder.sum() == 1
- thread C calls close() on root segment
    * sets state = CLOSING
    * calls adder.sum() which starts summing LongAdder cells and just misses adder.increment() of thread A2...
- thread A2 tries to acquire child segment and so it 1st does adder.increment()
- thread A2 reads state == CLOSING and therefore undoes previous increment with adder.decrement() and fails, but...
- thread C that just missed A2's adder.increment() does see A2's adder.decrement() because it is performed on a different LongAdder cell which C reads after it is incremented, so what C gets as a result of adder.sum() is 0, it completes closing the root segment by setting state = CLOSED and executing cleanup, but thread A1 still has access to the child segment and BOOM!!!

Having two LongAdders (acquires, releases) and doing the sum() per-partes sequentially: 1st releases.sum(), then acquires.sum() we guarantee that we never miss an acquiring thread's acquires.increment() but see the thread's releases.increment(), so above scenario is not possible.

Reply:
I agree with Peter that using a single long would misses update and you could see 'zero' when in reality something has happened but `sum` didn't get the memo. The entire approach, as far as I understand, relies on subtle happens-before relationship which arise precisely because of the use of the two counters (and the fact that one counter is always updated before the other).

Reply:
...right and the fact that the sum() of both counters is performed in the opposite order as they are updated.